blog post: vite support for ember-inspector (#2734)
mainmatter/mainmatter.com: src/posts/2025-06-20-ember-inspector-vite.md

Our team is relatively free to work on the ember-inspector part because [Chris Manson](https://mainmatter.com/blog/author/real_ate/) belongs to the [Ember Tooling Core Team](https://emberjs.com/teams/) and has permission to merge ready-to-go pull requests.

On the other hand, proposing changes in ember.js requires going through the [RFC process](https://rfcs.emberjs.com/). The RFC should describe the purpose and the accurate design of the new API that will expose the modules. Once written, the community will review and challenge it. Then, when a consensus is reached, the RFC gets the "accepted" state and can be implemented. This is a longer process by design, and we don't have full control over the timeline.

## Start with a proof of concept

To approach this work and draw the next steps, we started by implementing a proof of concept: We forked ember.js and ember-inspector and created testing branches that are not intended to be merged to design the new interaction system. Our approach relies on a global loading function exposed by ember-source and top-level `await` on the Inspector side to wait for the modules to be loaded.

> Out of our functional but rough proof of concept, we started to dig deeper into the Inspector side to refine the implementation and figure out all the pieces. By doing this first, we will kill two birds with one stone: we will prepare the ground for Vite support by managing any refactoring that turns out to be necessary, and we will find out the exact list of modules the ember-inspector relies on to reuse it in the future RFC.

Reviewer: ```suggestion
Out of our functional but rough proof of concept, we started to dig deeper into the Inspector side to refine the implementation and figure out all the pieces. By doing this first, we will kill two birds with one stone: we will prepare the ground for Vite support by managing any refactoring that turns out to be necessary, and we will find out the exact list of modules the ember-inspector relies on and can use this list to write the RFC.
```

- `3.16~3.24` scenarios were failing because of the structure of the tests in the repository. As you learned earlier in this blog post, the Inspector comprises a UI app and `ember_debug`. The piece we want to test against the ember-try scenarios is `ember_debug`: it's the piece that directly interacts with the inspected app, which can use any version of ember-source. The Inspector UI is just what it is, an Ember app using its own version of ember-source. The problem is that tests build the UI and `ember_debug` together, and when the scenarios run, modern syntax used on the UI side can trigger failure in the oldest versions, even though they are perfectly functional. This behavior has been patched by overriding the UI app in tests, but the tests' structure in the Inspector deserves to be rethought.

- `release`, `beta`, and `canary` (6.x) scenarios were failing essentially because the way ember-source exposes the modules changed. These versions introduce an `ember/barrel` module that the Inspector didn't know about. Additionally, non-colocated components are no longer allowed in these versions, so a few fixtures had to be rewritten in tests to adjust to this breaking change.

> An interesting part of this was the large contribution of [Patrick Pircher](https://github.com/patricklx). (Many thanks to him!) Sometimes, open source doesn't consist of coding things but rather of guiding others through a certain strategy and helping them help you.

Reviewer: I feel like we need to link to some of the PRs that patrick was doing rather than just to him directly 🤔 maybe have the sentence say "An interesting part of this was a series of large contributions from Patrick Pircher" and link to a search that has those PRs in it. What do you think? 

Also I think the "(Many thanks to him!)" should go before the full stop 🤔 since it's related to the sentence about the contributions

Embroider also has a concept of "adapter" that allows the transformation of v1 addons to make them compatible with Vite. This feature can be used to adapt the virtual content in Ember <= 6.1 when the path to modules changed (e.g. `@ember/enumerable/mutable` didn't exist before 4.8, and we should instead import from `@ember/-internals/runtime/lib/mixins/mutable_enumerable`).

We have already drafted a proof of concept to get a picture of what the implementation would look like.

### Watch and implement the RFC

The rest of the plan is straightforward: we will respond to the comments on the RFC as they come, and once the RFC is accepted, we will implement it in ember-source.

Since the review will take some time and won't require a full-time investment from us, we will parallelize watching the RFC with starting the next topic of the Ember Initiative: the router API.

## Summary

> Getting the Ember Inspector to support Vite apps is a demanding project that requires its own micro-roadmap, and involves three different repositories: ember-inspector, ember.js, and potentially Embroider for < 4.8 support. We have designed the plan, started to apply it, and made significant progress, overcoming hidden obstacles as they arise. We are still in the middle of the way, and the Ember Inspector should keep our team busy for a couple of weeks.

Reviewer: I don't understand what you're trying to say with "We are still in the middle of the way" 🤔 it doesn't sound right but since I don't fully understand so I can't give a suggestion 😂

Author: Sorry sometimes I just intuitively translate French expressions without realizing it’s an expression. I meant something like “we’ve done half of the work” or maybe “we’re still halfway there” except that “half” is really an approximation and not something super accurate. In French I would have said “on est encore au milieu du chemin” which is somehow more approximate than "on a fait la moitié du chemin".
